### PR TITLE
Introduce memory observer for ORTModule

### PR DESCRIPTION
**Pull request by pengwa**
### Introduce memory observer for ORTModule

To analyze memory usage for ORTModule training, we need collect per-iteration memory footprint in different stages (pre-forward, post-forward, pre-backward, and post-backward). 

Currently we only collect the data using torch.cuda APIs. The next step is, we could collect the detailed stashed activation list and its percentage within ORT backend, which is beyond this PR. 

Sample as below: 
```
0/8] step 0 memory (MiB) | phase: pre_forward | allocated: 1866 | max allocated: 1866 | cached: 1874 | max cached: 1874 | inactive: 8 | max inactive: 8
[0/8] step 0 memory (MiB) | phase: post_forward | allocated: 23277 | max allocated: 26215 | cached: 26406 | max cached: 26406 | inactive: 193 | max inactive: 405
[0/8] step 0 memory (MiB) | phase: pre_backward | allocated: 23277 | max allocated: 26215 | cached: 26406 | max cached: 26406 | inactive: 193 | max inactive: 405
[0/8] step 0 memory (MiB) | phase: post_backward | allocated: 2932 | max allocated: 26215 | cached: 26406 | max cached: 26406 | inactive: 6158 | max inactive: 6158
  0%|█                                                                                                                                                                                                            | 1/200 [00:26<1:26:18, 26.02s/it]
[0/8] step 1 memory (MiB) | phase: pre_forward | allocated: 2356 | max allocated: 26215 | cached: 26406 | max cached: 26406 | inactive: 2454 | max inactive: 6165
[0/8] step 1 memory (MiB) | phase: post_forward | allocated: 23767 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 2639 | max inactive: 6165
[0/8] step 1 memory (MiB) | phase: pre_backward | allocated: 23767 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 2639 | max inactive: 6165
[0/8] step 1 memory (MiB) | phase: post_backward | allocated: 3422 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 5284 | max inactive: 6165
  1%|██                                                                                                                                                                                                             | 2/200 [00:26<36:47, 11.15s/it]
[0/8] step 2 memory (MiB) | phase: pre_forward | allocated: 2356 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 2454 | max inactive: 6165
[0/8] step 2 memory (MiB) | phase: post_forward | allocated: 23767 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 2639 | max inactive: 6165
[0/8] step 2 memory (MiB) | phase: pre_backward | allocated: 23767 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 2639 | max inactive: 6165
[0/8] step 2 memory (MiB) | phase: post_backward | allocated: 3422 | max allocated: 26705 | cached: 29342 | max cached: 29342 | inactive: 5284 | max inactive: 6165
```

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py [hunk 1]
```diff
@@ -422,3 +465,87 @@ def _try_get_initializer_value(self, model, name):
...
+            ["max inactive", max_mem_inactive],  # peak of inactive, non-releasable memory
+        ]
+
+        summ = f"{self._rank_info} step {self._current_step} memory ({MemoryObserver.NORMALIZER_UNIT})"
+        for stat in mem_stats:
+            summ += f" | {stat[0]}: {stat[1]}"
+
+        # For the 10+ steps, only print when it is power of 2.
+        if self._current_step < 10 or (self._current_step & (self._current_step - 1) == 0):
+            print(summ)
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py:538]: can we use logger to log this?
something like this: 
```suggestion
            self._logger.info(summ)
```

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py [hunk 2]
```diff
@@ -422,3 +465,87 @@ def _try_get_initializer_value(self, model, name):
...
+            self._increase_step()
+            self._pre_phase = Phase.INVALID
+            return
+
+        self._pre_phase = cur_phase
+
+    def _increase_step(self):
+        self._current_step += 1
+
+    def _normalize(self, mem_size_in_bytes: Union[float, int]):
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py:550]:
```suggestion
    def _normalize(self, mem_size_in_bytes: Union[float, int]) -> str:
```

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 3]
```diff
@@ -190,8 +196,11 @@ def backward(ctx, *grad_outputs):
 
                 # Fast version: all backward_outputs are converted first.
                 # This version only works if backward_outputs is an OrtValueVector.
-                transfered_backward_outputs = _utils._ortvalues_to_torch_tensor(backward_outputs, self._device)
-                return tuple(transfered_backward_outputs[idx] if idx != -1 else None for idx in self._gradient_map)
+                transferred_backward_outputs = _utils._ortvalues_to_torch_tensor(backward_outputs, self._device)
+
+                self._rt_inspector.inspect_memory(Phase.POST_BACKWARD)
+
+                return tuple(transferred_backward_outputs[idx] if idx != -1 else None for idx in self._gradient_map)
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:203]: typo here

```suggestion
                transfered_backward_outputs = _utils._ortvalues_to_torch_tensor(backward_outputs, self._device)

                self._rt_inspector.inspect_memory(Phase.POST_BACKWARD)

                return tuple(transfered_backward_outputs [idx] if idx != -1 else None for idx in self._gradient_map)
```

Comment (mindest) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:203]: The repeated "r" should be correct.

**Review by mindest — COMMENTED** [review]
some typos

File: docs/ORTModule_Training_Guidelines.md [hunk 4]
```diff
@@ -158,6 +158,17 @@ inspection results to standard outputs.
 	export ORTMODULE_PRINT_INPUT_DENSITY=0 # Disable
 	```
 
+#### ORTMODULE_PRINT_MEMORY_STATS
+
+- **Feature Area**: *ORTMODULE/RuntimeInspector*
+- **Description**: By default, this is disabled. This env var can be used for print the memory inspection results
```
[the remark this answers is not in the document]
Comment (pengwa) [reply docs/ORTModule_Training_Guidelines.md]: Good catch, thanks man!

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py]: Good catch! Thanks!

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: Ah, sorry, guys. My mistake. :)

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: OK, not a mistake. transferred_backward_outputs  is correct one. I will revert it back.

**Review by AdamLouly — APPROVED** [review]
LGTM

**Review by mindest — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks @mindest, @AdamLouly for the review.

**Pull request merged**